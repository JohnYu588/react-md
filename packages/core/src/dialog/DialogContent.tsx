import type { HTMLAttributes } from "react";
import { forwardRef } from "react";
import type { DialogContentClassNameOptions } from "./styles";
import { dialogContent } from "./styles";

export interface DialogContentProps
  extends HTMLAttributes<HTMLDivElement>,
    DialogContentClassNameOptions {}

/**
 * **Server Component**
 *
 * Look at the `Dialog` or `FixedDialog` components for example usage.
 */
export const DialogContent = forwardRef<HTMLDivElement, DialogContentProps>(
  function DialogContent(
    { children, className, disablePadding = false, ...props },
    ref
  ) {
    return (
      <div
        {...props}
        ref={ref}
        className={dialogContent({
          className,
          disablePadding,
        })}
      >
        {children}
      </div>
    );
  }
);
