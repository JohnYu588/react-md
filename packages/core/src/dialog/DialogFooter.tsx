import { type HTMLAttributes, forwardRef } from "react";

import { type DialogFooterClassNameOptions, dialogFooter } from "./styles.js";

export interface DialogFooterProps
  extends HTMLAttributes<HTMLDivElement>,
    DialogFooterClassNameOptions {}

/**
 * The `DialogFooter` is a simple `<footer>` with simple `display: flex` styles
 * applied. Look at the `Dialog` or `FixedDialog` components for example usage.
 *
 * @see {@link https://next.react-md.dev/components/dialog|Dialog Demos}
 */
export const DialogFooter = forwardRef<HTMLDivElement, DialogFooterProps>(
  function DialogFooter({ children, className, align = "end", ...props }, ref) {
    return (
      <footer
        {...props}
        ref={ref}
        className={dialogFooter({
          align,
          className,
        })}
      >
        {children}
      </footer>
    );
  }
);
