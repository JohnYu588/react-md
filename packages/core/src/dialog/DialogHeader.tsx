import { forwardRef } from "react";

import { Box, type BoxProps } from "../box/Box.js";
import { type BoxAlignItems } from "../box/styles.js";
import { dialogHeader } from "./styles.js";

/**
 * @since 6.0.0 Extends the `BoxProps`.
 */
export interface DialogHeaderProps extends BoxProps {
  /** @defaultValue `"center"` */
  align?: BoxAlignItems;

  /** @defaultValue `true` */
  disableWrap?: boolean;
}

/**
 * The `DialogHeader` component should be rendered within a `Dialog` component
 * and generally contains the `DialogTitle`. Look at the `Dialog` or
 * `FixedDialog` components for example usage.
 *
 * @see {@link https://next.react-md.dev/components/dialog | Dialog Demos}
 * @since 6.0.0 Extends the `Box` component.
 */
export const DialogHeader = forwardRef<HTMLDivElement, DialogHeaderProps>(
  function DialogHeader(props, ref) {
    const {
      align = "center",
      disableWrap = true,
      children,
      className,
      ...remaining
    } = props;

    return (
      <Box
        {...remaining}
        align={align}
        disableWrap={disableWrap}
        ref={ref}
        className={dialogHeader({ className })}
      >
        {children}
      </Box>
    );
  }
);
