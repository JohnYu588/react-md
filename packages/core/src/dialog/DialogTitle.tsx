import { cnb } from "cnbuilder";
import { forwardRef } from "react";
import type {
  CustomTypographyComponent,
  TypographyMargin,
  TypographyProps,
  TypographyType,
} from "../typography/Typography.js";
import { Typography } from "../typography/Typography.js";

/**
 * @remarks \@since 6.0.0 Inherits the `TypographyProps` instead of
 * `HTMLAttributes<HTMLHeadingElement>`.
 */
export interface DialogTitleProps extends TypographyProps {
  /** @defaultValue `"h2"` */
  as?: CustomTypographyComponent;

  /** @defaultValue `"headline-4"` */
  type?: TypographyType;

  /** @defaultValue `"none"` */
  margin?: TypographyMargin;
}

/**
 * **Server Component**
 *
 * Look at the `Dialog` or `FixedDialog` components for example usage.
 */
export const DialogTitle = forwardRef<HTMLHeadingElement, DialogTitleProps>(
  function DialogTitle(
    {
      as = "h2",
      type = "headline-4",
      margin = "none",
      children,
      className,
      ...props
    },
    ref
  ) {
    return (
      <Typography
        {...props}
        ref={ref}
        as={as}
        type={type}
        margin={margin}
        // Note: This class applies no styles at this time
        className={cnb("rmd-dialog__title", className)}
      >
        {children}
      </Typography>
    );
  }
);
