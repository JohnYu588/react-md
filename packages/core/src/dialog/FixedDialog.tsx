"use client";

import { type RefObject, forwardRef, useRef } from "react";

import { TOP_INNER_RIGHT_ANCHOR } from "../positioning/constants.js";
import {
  type CalculateFixedPositionOptions,
  type PositionAnchor,
} from "../positioning/types.js";
import { useFixedPositioning } from "../positioning/useFixedPositioning.js";
import { DEFAULT_SCALE_CLASSNAMES } from "../transition/useScaleTransition.js";
import { type LabelRequiredForA11y } from "../types.js";
import { type BaseDialogProps, Dialog } from "./Dialog.js";

export interface BaseFixedDialogProps extends Omit<BaseDialogProps, "type"> {
  anchor?: PositionAnchor;
  fixedTo: RefObject<HTMLElement>;
  options?: CalculateFixedPositionOptions;
  getFixedPositionOptions?: () => CalculateFixedPositionOptions;

  /**
   * @defaultValue `true`
   */
  overlayHidden?: boolean;

  /**
   * @defaultValue `true`
   */
  disableScrollLock?: boolean;
}

export type FixedDialogProps = LabelRequiredForA11y<BaseFixedDialogProps>;

const noop = (): void => {
  // do nothing
};

/**
 * **Client Component**
 *
 * @example Simple Example
 * ```tsx
 * import { Button } from "@react-md/core/button/Button";
 * import { Dialog } from "@react-md/core/dialog/Dialog";
 * import { DialogContent } from "@react-md/core/dialog/DialogContent";
 * import { DialogFooter } from "@react-md/core/dialog/DialogFooter";
 * import { DialogHeader } from "@react-md/core/dialog/DialogHeader";
 * import { DialogTitle } from "@react-md/core/dialog/DialogTitle";
 * import { FixedDialog } from "@react-md/core/dialog/FixedDialog";
 * import { Typography } from "@react-md/core/typography/Typography";
 * import { useToggle } from "@react-md/core/useToggle";
 * import { useId, useRef, type ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   const {
 *     toggle,
 *     disable: onRequestClose,
 *     toggled: visible,
 *   } = useToggle(false);
 *   const titleId = useId();
 *   const fixedTo = useRef<HTMLButtonElement>(null)
 *
 *   return (
 *     <>
 *       <Button ref={fixedTo} onClick={toggle}>Toggle</Button>
 *       <FixedDialog
 *         aria-labelledby={titleId}
 *         fixedTo={fixedTo}
 *         visible={visible}
 *         onRequestClose={onRequestClose}
 *       >
 *         <DialogHeader>
 *           <DialogTitle id={titleId}>Simple Dialog</DialogTitle>
 *         </DialogHeader>
 *         <DialogContent>
 *           <Typography margin="none">This is some text in a dialog.</Typography>
 *         </DialogContent>
 *         <DialogFooter>
 *           <Button onClick={onRequestClose}>
 *             Close
 *           </Button>
 *         </DialogFooter>
 *       </FixedDialog>
 *     </>
 *   );
 * }
 * ```
 */
export const FixedDialog = forwardRef<HTMLDivElement, FixedDialogProps>(
  function FixedDialog(props, nodeRef) {
    const {
      fixedTo,
      style: propStyle,
      classNames = DEFAULT_SCALE_CLASSNAMES,
      children,
      anchor = TOP_INNER_RIGHT_ANCHOR,
      options,
      getFixedPositionOptions,
      onEnter = noop,
      onEntering,
      onEntered = noop,
      onExited,
      onRequestClose,
      overlayHidden = true,
      disableScrollLock = true,
      isFocusTypeDisabled = noop,
      ...remaining
    } = props;

    const disableExitFocus = useRef(false);
    const { ref, style, callbacks } = useFixedPositioning({
      nodeRef,
      style: propStyle,
      transformOrigin: true,
      onEnter(appearing) {
        onEnter(appearing);
        disableExitFocus.current = false;
      },
      onEntering,
      onEntered(appearing) {
        onEntered(appearing);
        disableExitFocus.current = false;
      },
      onExited,
      anchor,
      fixedTo,
      onScroll(_event, data) {
        if (!data.visible) {
          disableExitFocus.current = true;
          onRequestClose();
        }
      },
      ...options,
      getFixedPositionOptions,
    });

    return (
      <Dialog
        {...remaining}
        {...callbacks}
        ref={ref}
        type="custom"
        fixed
        style={style}
        classNames={classNames}
        onRequestClose={onRequestClose}
        overlayHidden={overlayHidden}
        disableScrollLock={disableScrollLock}
        isFocusTypeDisabled={(type) =>
          isFocusTypeDisabled(type) ||
          (type === "unmount" && disableExitFocus.current)
        }
      >
        {children}
      </Dialog>
    );
  }
);
