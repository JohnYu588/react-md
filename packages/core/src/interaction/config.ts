import { type ElementInteractionMode } from "./types.js";

/**
 * @remarks \@since 6.0.0
 */
export interface ElementInteractionConfiguration {
  mode: ElementInteractionMode;
  higherContrast: boolean;
}

// NOTE: Uses get/set for test mocking

let mode: ElementInteractionMode = "ripple";
let higherContrast = true;

/**
 * @remarks \@since 6.0.0
 */
export const INTERACTION_CONFIG: ElementInteractionConfiguration = {
  get mode() {
    return mode;
  },
  set mode(nextMode: ElementInteractionMode) {
    mode = nextMode;
  },
  get higherContrast() {
    return higherContrast;
  },
  set higherContrast(nextHigherContrast) {
    higherContrast = nextHigherContrast;
  },
};
