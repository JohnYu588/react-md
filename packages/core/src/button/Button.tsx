"use client";
import { cnb } from "cnbuilder";
import type { ButtonHTMLAttributes } from "react";
import { forwardRef } from "react";
import { RippleContainer } from "../interaction/RippleContainer";
import { useElementInteraction } from "../interaction/useElementInteraction";
import { useHigherContrastChildren } from "../interaction/useHigherContrastChildren";
import type { PropsWithRef } from "../types";
import { bem } from "../utils";
import type {
  FloatingActionButtonPosition,
  FloatingActionButtonProps,
} from "./FloatingActionButton";
import { FloatingActionButton } from "./FloatingActionButton";

declare module "react" {
  interface CSSProperties {
    "--rmd-button-background-color"?: string;
    "--rmd-button-color"?: string;
    "--rmd-button-border-radius"?: string | number;
    "--rmd-button-contained-background-color"?: string;
    "--rmd-button-contained-color"?: string;
    "--rmd-button-text-horizontal-padding"?: string | number;
    "--rmd-button-text-vertical-padding"?: string | number;
    "--rmd-button-text-min-height"?: string | number;
    "--rmd-button-text-min-width"?: string | number;
    "--rmd-button-icon-size"?: string | number;
    "--rmd-button-icon-radius"?: string | number;
  }
}

const styles = bem("rmd-button");

/**
 * When this is set to `"text"`, the size of the button will be determined by
 * the content and will be more block-like. Icons can still be rendered in
 * `"text"` buttons and will have spacing automatically applied between other
 * content in the button.
 *
 * When this is set to `"icon"`, the button will be equal height/width and
 * circular.
 */
export type ButtonType = "text" | "icon" | "icon-square";

/**
 * One of the valid material design default button themes that can be used. This
 * will update the general look and feel by updating the colors within the
 * button while the `ButtonThemeType` will update the borders or box shadow.
 */
export type ButtonTheme =
  | "clear"
  | "primary"
  | "secondary"
  | "warning"
  | "error"
  | "disabled";

/**
 * When this is set to `"flat"`, the button will have no `background-color`,
 * `border`, and `box-shadow`. It will only set the `color` to the
 * {@link ButtonTheme}.
 *
 * When this is set to `"outline"`, the button will have no `background-color`,
 * but gain a `border` and `color` set to the {@link ButtonTheme}.
 *
 * When this is set to `"contained"`, the button will set the `background-color`
 * to the {@link ButtonTheme}, add some `box-shadow`, and set the `color` to
 * either `#000` or `#fff`. (The `color` defaults to whichever value has the
 * highest contrast ratio with the `background-color`)
 */
export type ButtonThemeType = "flat" | "outline" | "contained";

/** @remarks \@since 6.0.0 */
export interface ButtonClassNameThemeOptions {
  className?: string;

  /** @defaultValue `false` */
  disabled?: boolean;

  /**
   * @see {@link ButtonTheme} for information about the different types.
   * @defaultValue `"text"`
   */
  buttonType?: ButtonType;

  /** @defaultValue `"clear"` */
  theme?: ButtonTheme;

  /**
   * @see {@link ButtonThemeType} for information about the theming behavior.
   * @defaultValue `"flat"`
   */
  themeType?: ButtonThemeType;
}

/**
 * @remarks \@since 6.0.0
 * @internal
 */
export interface ButtonClassNameOptions extends ButtonClassNameThemeOptions {
  /** @defaultValue `false` */
  pressed?: boolean;
  pressedClassName?: string;
}

/**
 * Creates a button theme based on the button theming props. This is really just
 * used so that other elements like clickable `<div>`s or `<input type="file">`
 * can look like buttons.
 *
 * @param options - An object containing the themeable button props to generate a
 * button theme className.
 * @returns a string of class names to create an element with a button theme.
 * @remarks \@since 6.0.0 This used to be called `buttonThemeClassNames`.
 */
export function button(options: ButtonClassNameOptions = {}): string {
  const {
    theme: propTheme = "clear",
    themeType = "flat",
    buttonType = "text",
    disabled: propDisabled = false,
    pressed = false,
    pressedClassName,
    className,
  } = options;

  const theme = propTheme === "disabled" ? "clear" : propTheme;
  const disabled = propDisabled || propTheme === "disabled";
  const text = buttonType === "text";
  const icon = !text;
  const outline = themeType === "outline";
  const contained = themeType === "contained";
  const clear = theme === "clear";

  return cnb(
    styles({
      text,
      icon,
      "icon-square": buttonType === "icon-square",
      disabled,
      contained: !disabled && contained,
      outline,
      pressed: contained && pressed,
      [theme]: !disabled && !clear && contained,
      [`text-${theme}`]: !disabled && !clear && !contained,
      [`outline-${theme}`]: !disabled && !clear && outline,
    }),
    pressedClassName,
    className
  );
}

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ButtonClassNameThemeOptions {
  /** @defaultValue `"button"` */
  type?: "button" | "reset" | "submit";

  /**
   * The position within the viewport to display the button as a floating action
   * button.
   */
  floating?: FloatingActionButtonPosition;

  /**
   * Any additional props to provide the to `FAB` container element when the
   * `floating` prop is provided
   */
  floatingProps?: PropsWithRef<FloatingActionButtonProps, HTMLSpanElement>;
}

/**
 * **Client Component**
 *
 * @example
 * Simple Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Theme Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="primary"
 *       themeType="contained"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Icon Button Example
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       theme="secondary"
 *       themeType="outline"
 *       buttonType="icon"
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *     </Button>
 *   );
 * }
 * ```
 *
 * @example
 * Text Button with icons
 * ```tsx
 * import { Button } from "@react-md/core";
 * import { FavoriteSVGIcon } from "@react-md/material-icons";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <Button
 *       onClick={(event) => {
 *         // do something
 *       }}
 *     >
 *       <FavoriteSVGIcon />
 *       Content
 *     </Button>
 *   );
 * }
 * ```
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(props, ref) {
    const {
      type = "button",
      disabled = false,
      floating = null,
      floatingProps,
      theme = floating ? "secondary" : "clear",
      themeType = floating ? "contained" : "flat",
      buttonType = floating ? "icon" : "text",
      className,
      children: propChildren,
      onBlur,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onDragStart,
      onTouchStart,
      onTouchEnd,
      onTouchMove,
      ...remaining
    } = props;
    const isThemeDisabled = theme === "disabled";
    const { pressed, pressedClassName, rippleContainerProps, handlers } =
      useElementInteraction({
        onBlur,
        onClick,
        onKeyDown,
        onKeyUp,
        onMouseDown,
        onMouseUp,
        onMouseLeave,
        onDragStart,
        onTouchStart,
        onTouchEnd,
        onTouchMove,
        disabled: disabled || isThemeDisabled,
      });

    const children = useHigherContrastChildren(propChildren);

    return (
      <FloatingActionButton position={floating} {...floatingProps}>
        <button
          {...remaining}
          // when the theme is set to `"disabled"`, the event handlers should be
          // removed so that it behaves like a disabled button. you do not want to
          // actually set the `disabled` attribute since it will lose keyboard
          // focus. this is mostly for supporting circular progress bars within
          // buttons
          {...(isThemeDisabled ? {} : handlers)}
          aria-disabled={isThemeDisabled || remaining["aria-disabled"]}
          disabled={disabled}
          ref={ref}
          type={type}
          className={button({
            theme,
            themeType,
            buttonType,
            disabled,
            pressed,
            pressedClassName,
            className,
          })}
        >
          {children}
          {rippleContainerProps && (
            <RippleContainer {...rippleContainerProps} />
          )}
        </button>
      </FloatingActionButton>
    );
  }
);
