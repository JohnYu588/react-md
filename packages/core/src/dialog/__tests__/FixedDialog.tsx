import { afterEach, describe, expect, it, jest } from "@jest/globals";
import { type ReactElement, type Ref, createRef, useId, useRef } from "react";

import { Button } from "../../button/Button.js";
import {
  act,
  render,
  rmdRender,
  screen,
  userEvent,
  waitFor,
} from "../../test-utils/index.js";
import { testImmediateRaf } from "../../test-utils/jest-globals/timers.js";
import { TRANSITION_CONFIG } from "../../transition/config.js";
import { SCALE_CLASSNAMES } from "../../transition/useScaleTransition.js";
import { Typography } from "../../typography/Typography.js";
import { useToggle } from "../../useToggle.js";
import { DialogContent } from "../DialogContent.js";
import { DialogFooter } from "../DialogFooter.js";
import { DialogHeader } from "../DialogHeader.js";
import { DialogTitle } from "../DialogTitle.js";
import { FixedDialog, type FixedDialogProps } from "../FixedDialog.js";

type TestProps = Omit<
  FixedDialogProps,
  "fixedTo" | "visible" | "onRequestClose"
> & {
  nodeRef?: Ref<HTMLDivElement>;
  defaultVisible?: boolean;
};

function Test(props: TestProps): ReactElement {
  const { defaultVisible = false, nodeRef, ...remaining } = props;
  const fixedTo = useRef<HTMLButtonElement>(null);
  const {
    toggled: visible,
    enable: showDialog,
    disable: hideDialog,
  } = useToggle(defaultVisible);
  const titleId = useId();
  return (
    <>
      <Button ref={fixedTo} onClick={showDialog}>
        Button
      </Button>
      <FixedDialog
        {...remaining}
        aria-labelledby={titleId}
        ref={nodeRef}
        fixedTo={fixedTo}
        visible={visible}
        onRequestClose={hideDialog}
      >
        <DialogHeader>
          <DialogTitle id={titleId}>Hello, world!</DialogTitle>
        </DialogHeader>
        <DialogContent>
          <Typography margin="none">Additional content</Typography>
        </DialogContent>
        <DialogFooter>
          <Button onClick={hideDialog}>OK</Button>
        </DialogFooter>
      </FixedDialog>
    </>
  );
}

describe("FixedDialog", () => {
  afterEach(() => {
    jest.useRealTimers();
    TRANSITION_CONFIG.disabled = true;
  });

  it("should apply the correct styling, HTMLAttributes, and allow a ref", () => {
    const ref = createRef<HTMLDivElement>();

    const { rerender } = render(<Test nodeRef={ref} defaultVisible />);
    const dialog = screen.getByRole("dialog", { name: "Hello, world!" });
    expect(ref.current).toBeInstanceOf(HTMLDivElement);
    expect(ref.current).toBe(dialog);
    expect(dialog).toMatchSnapshot();

    rerender(
      <Test
        nodeRef={ref}
        defaultVisible
        style={{ color: "orange" }}
        className="custom-class-name"
      />
    );
    expect(dialog).toHaveStyle("color: orange");
    expect(dialog).toHaveClass("custom-class-name");
    expect(dialog).toMatchSnapshot();
  });

  it("should default to hiding the overlay", async () => {
    const user = userEvent.setup();
    const { rerender } = rmdRender(
      <Test overlayProps={{ "data-testid": "overlay" }} />
    );

    expect(() => screen.getByRole("dialog")).toThrow();
    expect(() => screen.getByTestId("overlay")).toThrow();
    const button = screen.getByRole("button", { name: "Button" });
    await user.click(button);

    const overlay = screen.getByTestId("overlay");
    expect(overlay).not.toHaveClass("rmd-overlay--active");

    rerender(
      <Test overlayProps={{ "data-testid": "overlay" }} overlayHidden={false} />
    );
    expect(overlay).toHaveClass("rmd-overlay--active");
  });

  it("should default to using the SCALE_CLASSNAMES over the DEFAULT_DIALOG_CLASSNAMES for the transition", async () => {
    jest.useFakeTimers();
    TRANSITION_CONFIG.disabled = false;

    const user = userEvent.setup({
      delay: null,
    });
    rmdRender(<Test overlayProps={{ "data-testid": "overlay" }} />);

    await user.click(screen.getByRole("button", { name: "Button" }));
    const dialog = screen.getByRole("dialog", { name: "Hello, world!" });
    expect(dialog).toHaveClass(SCALE_CLASSNAMES.enter);
    expect(dialog).toHaveClass(SCALE_CLASSNAMES.enterActive);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(dialog).not.toHaveClass(SCALE_CLASSNAMES.enter);
    expect(dialog).not.toHaveClass(SCALE_CLASSNAMES.enterActive);

    act(() => {
      jest.runOnlyPendingTimers();
    });
    await user.click(screen.getByTestId("overlay"));
    expect(dialog).toHaveClass(SCALE_CLASSNAMES.exit);
    expect(dialog).toHaveClass(SCALE_CLASSNAMES.exitActive);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(dialog).not.toBeInTheDocument();
  });

  it("should default to allowing the user to scroll the page while the dialog is visible", () => {
    const { rerender } = render(<Test defaultVisible />);
    expect(document.body).not.toHaveStyle("overflow: hidden");

    rerender(<Test defaultVisible disableScrollLock={false} />);
    expect(document.body).toHaveStyle("overflow: hidden");
  });

  it("should prevent refocusing the fixedTo element when it is no longer visible in the viewport to prevent scrolling back to that element", async () => {
    const user = userEvent.setup();
    const raf = testImmediateRaf();
    rmdRender(<Test />);
    const button = screen.getByRole("button", { name: "Button" });

    await user.click(button);
    let dialog = screen.getByRole("dialog", { name: "Hello, world!" });
    expect(dialog).toHaveFocus();

    await user.keyboard("[Escape]");
    expect(dialog).not.toBeInTheDocument();
    await waitFor(() => {
      expect(button).toHaveFocus();
    });

    await user.click(button);
    dialog = screen.getByRole("dialog", { name: "Hello, world!" });
    expect(dialog).toHaveFocus();

    jest.spyOn(button, "getBoundingClientRect").mockReturnValue({
      ...document.body.getBoundingClientRect(),
      top: -300,
      bottom: -200,
    });
    jest.spyOn(dialog, "getBoundingClientRect").mockReturnValueOnce({
      ...document.body.getBoundingClientRect(),
      top: -280,
      bottom: -100,
    });
    act(() => {
      const scrollEvent = new UIEvent("scroll");
      window.dispatchEvent(scrollEvent);
    });
    expect(dialog).not.toBeInTheDocument();
    expect(button).not.toHaveFocus();
    expect(document.body).toHaveFocus();

    raf.mockRestore();
  });
});
