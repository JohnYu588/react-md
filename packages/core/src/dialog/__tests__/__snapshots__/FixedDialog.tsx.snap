// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedDialog should apply the correct styling, HTMLAttributes, and allow a ref 1`] = `
<div
  aria-labelledby=":r0:"
  class="rmd-dialog rmd-dialog--fixed rmd-dialog--outline"
  id="dialog-:r1:"
  role="dialog"
  style="position: fixed; transform-origin: 0 0; left: 16px; top: 0px;"
  tabindex="-1"
>
  <div
    class="rmd-box rmd-box--gap rmd-box--padded rmd-box--align-center rmd-dialog__header"
  >
    <h2
      class="rmd-typography rmd-typography--headline-4 rmd-no-margin rmd-dialog__title"
      id=":r0:"
    >
      Hello, world!
    </h2>
  </div>
  <div
    class="rmd-dialog__content rmd-dialog__content--padded"
  >
    <p
      class="rmd-typography rmd-typography--body-1 rmd-no-margin"
    >
      Additional content
    </p>
  </div>
  <footer
    class="rmd-dialog__footer rmd-dialog__footer--flex rmd-dialog__footer--end"
  >
    <button
      class="rmd-button rmd-button--text rmd-interaction-surface"
      type="button"
    >
      <span>
        OK
      </span>
    </button>
  </footer>
</div>
`;

exports[`FixedDialog should apply the correct styling, HTMLAttributes, and allow a ref 2`] = `
<div
  aria-labelledby=":r0:"
  class="rmd-dialog rmd-dialog--fixed rmd-dialog--outline custom-class-name"
  id="dialog-:r1:"
  role="dialog"
  style="position: fixed; transform-origin: 0 0; left: 16px; top: 0px; color: orange;"
  tabindex="-1"
>
  <div
    class="rmd-box rmd-box--gap rmd-box--padded rmd-box--align-center rmd-dialog__header"
  >
    <h2
      class="rmd-typography rmd-typography--headline-4 rmd-no-margin rmd-dialog__title"
      id=":r0:"
    >
      Hello, world!
    </h2>
  </div>
  <div
    class="rmd-dialog__content rmd-dialog__content--padded"
  >
    <p
      class="rmd-typography rmd-typography--body-1 rmd-no-margin"
    >
      Additional content
    </p>
  </div>
  <footer
    class="rmd-dialog__footer rmd-dialog__footer--flex rmd-dialog__footer--end"
  >
    <button
      class="rmd-button rmd-button--text rmd-interaction-surface"
      type="button"
    >
      <span>
        OK
      </span>
    </button>
  </footer>
</div>
`;
